packages/compass-e2e-tests/helpers/commands/run-find-operation.js

```diff
@@ -0,0 +1,129 @@
+const Selectors = require('../selectors');
+
+async function setFilter(client, tabName, value) {
+  await client.setAceValue(Selectors.queryBarOptionInputFilter(tabName), value);
+}
+
+async function setProject(client, tabName, value) {
+  await client.setAceValue(
+    Selectors.queryBarOptionInputProject(tabName),
+    value
+  );
+}
+
+async function setSort(client, tabName, value) {
+  await client.setAceValue(Selectors.queryBarOptionInputSort(tabName), value);
+}
+
+async function setCollation(client, tabName, value) {
+  await client.setAceValue(
+    Selectors.queryBarOptionInputCollation(tabName),
+    value
+  );
+}
+
+async function setMaxTimeMS(client, tabName, value) {
+  const selector = Selectors.queryBarOptionInputMaxTimeMS(tabName);
+  await client.clickVisible(selector);
+  await client.setValue(selector, value);
+}
+
+async function setSkip(client, tabName, value) {
+  const selector = Selectors.queryBarOptionInputSkip(tabName);
+  await client.clickVisible(selector);
+  await client.setValue(selector, value);
+}
+
+async function setLimit(client, tabName, value) {
+  const selector = Selectors.queryBarOptionInputLimit(tabName);
+  await client.clickVisible(selector);
+  await client.setValue(selector, value);
+}
+
+async function runFind(client, tabName) {
+  await client.clickVisible(Selectors.queryBarApplyFilterButton(tabName));
+}
+
+async function isOptionsExpanded(client, tabName) {
+  // it doesn't look like there's some attribute on the options button or container that we can easily check, so just look for a field that exists if it is expanded
+  return await client.isVisible(Selectors.queryBarOptionInputProject(tabName));
+}
+
+async function waitUntilCollapsed(client, tabName) {
+  await client.waitForVisible(
+    Selectors.queryBarOptionInputProject(tabName),
+    1000,
+    true
+  );
+}
+
+async function collapseOptions(client, tabName) {
+  if (!(await isOptionsExpanded(client, tabName))) {
+    return;
+  }
+
+  await client.clickVisible(Selectors.queryBarOptionsToggle(tabName));
+  await waitUntilCollapsed(client, tabName);
+}
+
+async function waitUntilExpanded(client, tabName) {
+  await client.waitForVisible(Selectors.queryBarOptionInputProject(tabName));
+}
+
+async function expandOptions(client, tabName) {
+  if (await isOptionsExpanded(client, tabName)) {
+    return;
+  }
+
+  await client.clickVisible(Selectors.queryBarOptionsToggle(tabName));
+  await waitUntilExpanded(client, tabName);
+}
+
+module.exports = function (app) {
+  return async function runFindOperation(
+    tabName,
+    filter,
+    {
+      project = '',
+      sort = '',
+      maxTimeMS = '',
+      collation = '',
+      skip = '',
+      limit = '',
+    } = {}
+  ) {
+    const { client } = app;
+
+    const queryBarSelector = Selectors.queryBar(tabName);
+
+    // look up the current resultId
+    const initialResultId = await client.getAttribute(
```

Comment:
This is basically what makes this command work with any of the three query bars (Documents, Schema or Explain Plan tabs). Whenever we run a query or reset it and it completes (either successfully or unsuccessfully) it will change this attribute to a new large random number and that way we can know that the query completed.

I considered waiting for that gray out thing and for the button to become active again, but the problem is that it happens to quickly and all the waitUntil type commands poll. Plus the end state is identical to the start state so you can't tell if it has already happened vs whether it will still happen.

From past experience I know that waiting for a thing that will go away by itself is a giant source of flakes in e2e tests - you have to wait for a stable state that won't change until the user (ie. the test) interacts with the page. ie. waiting or a loading spinner to appear/disappear is just never a good idea.

I also started out with looking for the page to update, but I needed code for each of the three tabs. And it quickly devolved into hacks like "search for something that can never match, wait for the text to say nothing matches, now run the actual query and wait for that text to change.."

So it kinda sucks that I had to change the app a bit to make it more testable, but sometimes that's the best way forward.